Support sdbus clients

mardy opened the pull request:
This is built on top of https://github.com/flatpak/xdg-dbus-proxy/pull/26. I've kept the original commit by @refi64 and added one to address the [issue raised during the review](https://github.com/flatpak/xdg-dbus-proxy/pull/26#issuecomment-1116071581).

Whether my commit actually succeeds in addressing that problem is something I'm not sure of, since I couldn't actually reproduce the original issue. But from my first tests it appears that sdbus clients can connect and send methods calls.

bilelmoussaoui commented on flatpak-proxy.c:
```diff
@@ -2959,7 +2997,49 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
+
+              got_buffer_from_side (side, buffer);
+
+              check_pending_auth_lines(client);
```
```suggestion
              check_pending_auth_lines (client);
```

bilelmoussaoui reviewed (commented):
various code style comments

sophie-h commented:
I cannot confirm that the patch fixes the issue for me.

mardy commented on flatpak-proxy.c:
```diff
@@ -2598,10 +2611,23 @@ got_buffer_from_client (FlatpakProxyClient *client, ProxySide *side, Buffer *buf
     queue_initial_name_ops (client);
 }
 
+static void check_pending_auth_lines(FlatpakProxyClient *client)
```
Updated.

sophie-h commented:
> You can use this flatpak https://github.com/zecakeh/zbus-flatpak-portal

If you just use `cargo run` outside a Flatpak it should run properly. Afterwards, you can run it from a flatpak by 

```
flatpak run --filesystem=host --command=./target/debug/zbus-flatpak-portal app.drey.Warp
```

Where you can use any id of an installed app instead of app.drey.Warp.

This then gives

```
Creating connection…
thread 'main' panicked at src/main.rs:7:60:
called `Result::unwrap()` on an `Err` value: InputOutput(Custom { kind: BrokenPipe, error: "failed to read from socket" })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

(Sorry if this is all obvious for you. Just wanted to make sure you can reproduce it.)

bilelmoussaoui commented:
> > You can use this flatpak https://github.com/zecakeh/zbus-flatpak-portal
> 
> If you just use `cargo run` outside a Flatpak it should run properly. Afterwards, you can run it from a flatpak by
> 
> ```
> flatpak run --filesystem=host --command=./target/debug/zbus-flatpak-portal app.drey.Warp
> ```
> 
> Where you can use any id of an installed app instead of app.drey.Warp.
> 
> This then gives
> 
> ```
> Creating connection…
> thread 'main' panicked at src/main.rs:7:60:
> called `Result::unwrap()` on an `Err` value: InputOutput(Custom { kind: BrokenPipe, error: "failed to read from socket" })
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> (Sorry if this is all obvious for you. Just wanted to make sure you can reproduce it.)

Note that zbus 4.2.1 includes a hack to workaround this issue, so you would have to make sure it is using zbus 4.2.0

mardy commented:
Thanks for your help! I added one more commit (I guess it can be squashed with the previous ones, but for now I'll keep it separate as the commit message sheds some light on it). Now I can run the `zbus-flatpak-portal` program.

I added one more member (`sent`) to the Buffer structure. It may be possible to do without it, but the code was hard to understand because the `pos` and `size` members where being used with a different meaning depending on whether we were reading from or writing to the buffer; and the fact that we pass buffers from one socket to the other increases on this confusion.

I didn't want to make a too invasive change, but if you agree I'd rename the members like this (suggestions welcome!):
- size -> capacity ("size" can also be OK, if you insist :-) )
- pos -> "length", "used", "data_size"?
- sent -> keep as is?

sophie-h reviewed (changes requested):
The latest version does seem to fix the problems for me!

@mardy could you maybe address the small review comments that are open such that we can hopefully move forward with this soon? :)

sophie-h commented:
I didn't want to make a too invasive change, but if you agree I'd rename the members like this (suggestions welcome!):

- size -> capacity ("size" can also be OK, if you insist :-) )
- pos -> "length", "used", "data_size"?
- sent -> keep as is?

Maintainers have been very open to making the code clearer. However, maybe you can add these changes as an additional commit to simplify the review?

swick commented on flatpak-proxy.c:
```diff
@@ -2959,7 +2982,48 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
-                client->authenticated = TRUE;
+                client->auth_state = client->auth_server_backlog
+                  ? AUTH_WAITING_FOR_BACKLOG
+                  : AUTH_COMPLETE;
```
Could use some braces.

swick replied:
And maybe be more explicit: `client->auth_server_backlog > 0`

mardy replied:
Mmm... This is the original commit from @refi64, which I based my work on. I included in the branch as-is and then I added my fixes on top of his.

I don't think I should change this commit. Maybe then it's better if I merge his and my commit (and I will fix all these remarks), and then I just credit him in the commit message? WDYT?

swick commented on flatpak-proxy.c:
```diff
@@ -2959,7 +2982,48 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
+                    {
+                      line_start = line_end + strlen (AUTH_LINE_SENTINEL);
+
+                      if (--client->auth_server_backlog == 0)
```
Not a fan of side effects in if statements.

swick commented on flatpak-proxy.c:
```diff
@@ -2959,7 +2982,48 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
+              got_buffer_from_side (side, buffer);
+
+              if (client->auth_server_backlog == 0)
+                client->auth_state = AUTH_COMPLETE;
```
Moving this to the end of the if above makes it easier to follow.

swick commented on flatpak-proxy.c:
```diff
@@ -298,6 +298,7 @@ struct FlatpakProxyClient
   AuthState     auth_state;
```
I don't understand what this commit changes. What is being fixed up exactly?

mardy commented:
> This contains commits which are broken by themselves. Can you please squash thing properly so that each commit is working?

By "broken" you mean "not building" or "not fixing the problem"? If it's the former, then I guess the problem is the warning/error [mentioned by Sophie](https://github.com/flatpak/xdg-dbus-proxy/pull/54#discussion_r1649694564) (which I will fix); if the latter, then I guess I should just squash all commits into one, because only the last commit fixes the issue for good.

mardy commented on flatpak-proxy.c:
```diff
@@ -2959,7 +3023,54 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
+                {
+                  /* Immediately send the outgoing buffer */
+                  GSocketConnection *connection = client->bus_side.connection;
+                  GSocket *socket = g_socket_connection_get_socket (connection);
```
Fixed.

alexlarsson commented on flatpak-proxy.c:
```diff
@@ -2940,7 +3004,54 @@ side_in_cb (GSocket *socket, GIOCondition condition, gpointer user_data)
+
+              got_buffer_from_side (side, buffer);
+
+              check_pending_auth_lines (client);
```
When we set AUTH_COMPLETE when we get a bus message there may be outstanding extra_input_data from the auth message on the client input  side, but unless we get some action on the input fd we will not read it. So we need to manually trigger a read in this case.

alexlarsson commented:
I pushed some lightly tested fixes for the issues I found. Please have a look at these.
Otherwise this looks good to me.

I'd like to not have any individual commits in the tree where things are broken, so that bisect works, so most commits here should imho be squashed into one commit. However, I made them separate for easier review.

swick commented on flatpak-proxy.c:
```diff
@@ -837,13 +857,18 @@ buffer_read (ProxySide *side,
+  if (side->extra_input_data && client->auth_state == AUTH_COMPLETE)
     {
       gsize extra_size;
       const guchar *extra_bytes = g_bytes_get_data (side->extra_input_data, &extra_size);
```
A few lines later:
```c
side->extra_input_data =
            g_bytes_new_with_free_func
```
This overwrites `side->extra_input_data` without freeing the previous one.

alexlarsson replied:
It frees the previous one by virtue of passing g_bytes_unref + side->extra_input_data as destroy notify.

swick replied:
Wow, that's some extremely subtle code here... have never seen this kind of cleanup chain. What's the reason here instead of just cleaning up the old one immediately?

alexlarsson replied:
Because the new GBytes references the memory of the first one. If we freed it early it would point into freed memory (or we would have to duplicate the memory).

alexlarsson replied:
Its basically:
` extra_input_data = extra_input_data[received:]`

alexlarsson replied:
I guess it would be cleaner if we always unref:ed it and explicitly ref:ed it in the re-use case.

swick replied:
Well, it's working and this MR isn't touching it, so let's just resolve this thread.

swick commented:
The changes look fine to me.

alexlarsson commented:
I found one more issue in testing, but now it seems to work, so I squashed the functional changes into one commit (with props to ryan and alberto).

I think this is fine to go, but please take a second look and test it.

swick commented:
Hard to see what changed because of  the indentation changes and rebase.

alexlarsson commented:
The only real change was to the handling of auth messages. We now delay the setting of AUTH_COMPLETE to after calling got_buffer_from_side() as that would otherwise try to parse the "BEGIN" auth line as a dbus message. See the new_auth_state variable.

But, it wouldn't be a bad idea to do a more complete review of the combined change, to ensure fresh eyes on this.

mardy commented:
> @mardy can you also please take another look?

Will do tomorrow :-)

mardy commented:
> > @mardy can you also please take another look?
> 
> Will do tomorrow :-)

I confirm that this continues to work for me :-)

alexlarsson commented:
merging then
